[feat] Add toggle to allow log lines to be JSON colorized

rm-hull opened the pull request:
I've got a tentative PR that adds colorized log lines when the line can be parsed as JSON (when it can't be parsed into JSON, the line is just returned as-is).

Colorizing can be applied permanently in the config, but by default is is toggled off, activated by the 'J' key:
<img width="1060" alt="Screenshot 2022-09-25 at 22 38 19" src="https://user-images.githubusercontent.com/1915543/192167145-0dd6b2f6-c8f3-41da-91c6-5735c8529173.png">

I've marked this as "Draft" for the moment because I'd like to add some tests over this behaviour, and I've also got an outstanding PR open against colorjson (see https://github.com/TylerBrock/colorjson/pull/7) which would be nice to get merged rather than using my forked version, cc/ @TylerBrock.

I followed how the log timestamps were implemented, and added colorization to the model rather than the view. Initially, I had added it to the view and this didn't play nicely with filter or timestamps (this is how I discovered and came to fix #1776), whereas it works fine when located in the model.

Anyhow, hopefully this would be a useful feature on an already amazingly useful tool!

derailed commented:
@rm-hull Very cool! This was a long awaited feature that many folks voiced. Thank you so much Richard for punching this thru and for the great explanation!!  That said, I think we have to look in a bit closer about the cost of this implementation. Having to marshal/unmarshal each line is bound to run gc bunkers especially in this already less than optimal implementation. Let me noodle on this a bit more and take your branch out for a spin...

rm-hull commented:
I didn't look too closely into the model code, but does it trigger the render for every log line or just the ones currently visible in the box view? If the former then, yes I imagine it could incur quite an overhead.

I did start off trying to write a naive parser that maintained a bit of state to determine the current color (whether in a string, in curly brackets, in square brackets, etc + managing that recursively) but it got complicated quickly.

Marshalling to JSON and the delegating to a lib to add color seemed like the right thing to do. 

If we can restrict the view to just rendering the visible log lines this might still be ok?

thinkniko commented:
@rm-hull Can you change the key to Shift-J for the json toggle?
When viewing logs, I use the J key for scrolling down after scrolling up with the K key.

rm-hull commented:
@thinkniko .. I changed the toggle to Ctrl-J

slimus commented on internal/dao/log_item.go:
```diff
@@ -89,9 +92,25 @@ func (l *LogItem) Render(paint string, showTime bool, bb *bytes.Buffer) {
+	if !showJson {
+		return text
+	}
+	var obj map[string]interface{}
```
What if you have int or object as an index, not a string?

rm-hull replied:
Json object keys must always be strings: https://www.w3schools.com/js/js_json_objects.asp#:~:text=Keys%20must%20be%20strings%2C%20and,number

slimus replied:
It's true if your application returns an object. But it can return an array, right?

slimus replied:
But nvm, I'm going to resolve this. Thanks!

slimus commented on internal/dao/log_item.go:
```diff
@@ -89,9 +92,25 @@ func (l *LogItem) Render(paint string, showTime bool, bb *bytes.Buffer) {
+
+	s, err := f.Marshal(obj)
+	if err != nil {
+		return text
```
I would like to see `err` in debug logs, but I don't know how... Here and at the line 105

rm-hull replied:
I can kind-of see why that might be useful, but if the text presented cannot be unmarshalled into a json object, then it’s not likely to be json. What more is there to say? Otherwise I fear it will just clog up the logs. Nevertheless I will have a look to see if it’s possible.

slimus replied:
I see only 1 use case: your application generates json-like log but with some problems. And in this case, you will see un-colored json and you can create an issue here, but the real problem is in your app :) Okay, I agree with you.

derailed commented:
@rm-hull Thanks for this PR Richard. Looks like we have conflicts. Could you update?

rm-hull commented:
> @rm-hull Thanks for this PR Richard. Looks like we have conflicts. Could you update?

Conflicts fixed

rm-hull commented:
Hi @derailed, @slimus - I rebased and fixed merge conflicts again, so the PR should be up-to-date w.r.t. the main branch .. would be good to get this in 🙏

KevinGimbel commented:
LGTM! I checked out the PR and tested it with a few pod logs, works as expected.

@derailed I think this can be merged with the next release.

froats commented:
up
